File: Binary Tree/二叉树的属性/257.二叉树的所有路径.cpp
#include <vector>
#include <string>

using namespace std;
/*
 * @lc app=leetcode.cn id=257 lang=cpp
 *
 * [257] 二叉树的所有路径
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    vector<string> result;
    void getPath(TreeNode *root, string rootVal)
    {
        if (rootVal == "")
        {
            rootVal = to_string(root->val);
        }
        else
        {
            rootVal = rootVal + "->" + to_string(root->val);
        }

        if (root->left != nullptr)
        {
            getPath(root->left, rootVal);
        }

        if (root->right != nullptr)
        {
            getPath(root->right, rootVal);
        }

        if (root->left == nullptr && root->right == nullptr)
        {
            result.push_back(rootVal);
        }
    }
    vector<string> binaryTreePaths(TreeNode *root)
    {
        if (root == nullptr)
        {
            return result;
        }

        getPath(root, "");

        return result;
    }
};
// @lc code=end

File: Binary Tree/二叉树的属性/257.二叉树的所有路径.ts
/*
 * @lc app=leetcode.cn id=257 lang=typescript
 *
 * [257] 二叉树的所有路径
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */
function getPath(str: string, root: TreeNode | null, result: any): void {
  if (!root) {
    return
  }

  if (!str) {
    str = root.val + ""
  } else {
    str += "->" + root.val
  }

  // 到达根节点，push进结果
  if (!root.left && !root.right) {
    result.push(str)
    return
  }

  getPath(str, root.left, result)
  getPath(str, root.right, result)
}

function binaryTreePaths(root: TreeNode | null): string[] {
  const result = []
  getPath("", root, result)
  return result
};
// @lc code=end

